lib/booking.rb

```diff
@@ -0,0 +1,242 @@
+
+require_relative 'exceptions'
+require 'pry'
+
+module Hotel
+  class Booking
+    attr_reader :all_reservations, :all_rooms, :all_blocks
+    def initialize
+      @all_reservations = []
+      @all_rooms = []
+      @all_blocks = []
+      (1..20).each do |number|
+        @all_rooms << Hotel::Room.new(number)
+      end
+    end # initialize
+
+    def valid_dates(start_date, end_date)
+      if end_date < start_date
+        raise BookingError.new("Your checkout day must be after your checkin date! You entered: checkin day = #{start_date} and checkout date = #{end_date}")
+      elsif end_date == start_date
+        raise BookingError.new("You must book at least one night. You entered: checkin day = #{start_date} and checkout date = #{end_date} ")
+      end # if
+    end # valid_dates
+
+    def get_date_array(start_date, end_date)
+      dates = Hotel::DateRange.new(start_date, end_date).nights_booked
+    end # get_date_array
+
+    def get_rooms(start_date, end_date, num)
+      availible = availible_rooms(start_date, end_date)
+      rooms = []
+      i = 0
+      num.times do
+        rooms << availible[i]
+        i += 1
+      end # .times
+      return rooms
+    end # get_rooms
+
+    def make_reservation(start_date, end_date, num_rooms )
+      valid_dates(start_date, end_date)
+      if start_date < end_date
+        availible = availible_rooms(start_date, end_date)
+        if num_rooms > 20
+          raise BookingError.new("You can't reserve that many rooms because we only have 20 rooms in the hotel.")
+        elsif num_rooms > availible.length
+          raise BookingError.new("We don't have that many rooms availible for those dates. We only have #{availible.length} rooms availible for those dates. ")
+        end # if/elsif
+        dates_booked = get_date_array(start_date, end_date)
+        reservation_id = @all_reservations.length + 1
+        rooms = get_rooms(start_date, end_date, num_rooms)
+        cost = (num_rooms * dates_booked.length * 200.0)
+        @all_reservations << Hotel::Reservation.new(reservation_id, cost, rooms, dates_booked)
+      end # if/else
+    end # make_reservation
+
+
+    def check_date_for_reservations(start_date, end_date)
+      # it will have a method that checks if there is a reservation for the date(s) requested
+      # iterate though @all_reservations for each day requested and if the day is included in the Booking then it will look at the Reservation that included that date and add the room(s) in that reservation to a new array not_availible.
+      days = get_date_array(start_date, end_date)
+      date_reservations = []
+      days.each do |day|
+        @all_reservations.each do |res|
+          if res.date_range.include?(day)
+            if !(date_reservations.include?(res))
```

Comment:
All the places that you create a `DateRange`, it seems like you're only using it as a tool to get an array of dates. This indicates that a method might be a better choice here, one that takes a checkin and checkout date and returns an array of dates.

The other option would be to write an instance method, something like `DateRange#overlap?(other)`, that takes another `DateRange` and checks to see if they overlap. Then this complex bit of logic would be hidden away in `DateRange` instead of sitting here in `Booking`.

As long as you've identified "manage a range of dates" as a responsibility, then it probably makes sense to do the later, and keep that responsibility out of the `Booking` class.

This would also make it much easier to test!